cql3/statements/cas_request.cc

```diff
@@ -120,21 +120,23 @@ std::optional<mutation> cas_request::apply(foreign_ptr<lw_shared_ptr<query::resu
 const update_parameters::prefetch_data::row* cas_request::find_old_row(const cas_row_update& op) const {
     static const clustering_key empty_ckey = clustering_key::make_empty();
     const partition_key& pkey = _key.front().start()->value().key().value();
-    // If a statement has only static columns conditions, we must ignore its clustering columns
-    // restriction when choosing a row to check the conditions, i.e. choose any partition row,
-    // because any of them must have static columns and that's all we need to know if the
-    // statement applies. For example, the following update must successfully apply (effectively
+    // We must ignore statement clustering column restriction when
+    // choosing a row to check the conditions. If there is no
+    // exact match, choose static row to check if the statement
+    // applies.
+    // For example, the following update must successfully apply (effectively
     // turn into INSERT), because, although the table doesn't have any regular rows matching the
     // statement clustering column restriction, the static row matches the statement condition:
     //   CREATE TABLE t(p int, c int, s int static, v int, PRIMARY KEY(p, c));
     //   INSERT INTO t(p, s) VALUES(1, 1);
     //   UPDATE t SET v=1 WHERE p=1 AND c=1 IF s=1;
-    // Another case when we pass an empty clustering key prefix is apparently when the table
-    // doesn't have any clustering key columns and the clustering key range is empty (open
-    // ended on both sides).
-    const clustering_key& ckey = !op.statement.has_only_static_column_conditions() && op.ranges.front().start() ?
-        op.ranges.front().start()->value() : empty_ckey;
-    return _rows.find_row(pkey, ckey);
+    const clustering_key& ckey = op.ranges.front().start() ?  op.ranges.front().start()->value() : empty_ckey;
+    auto row = _rows.find_row(pkey, ckey);
+    if (row == nullptr && !ckey.is_empty() &&
+        !op.statement.has_if_exist_condition() && !op.statement.has_if_not_exist_condition()) {
+        row = _rows.find_row(pkey, empty_ckey);
+    }
+    return row;
```

Comment:
I don't know if this is correct. I think the static-columns-only row materializes only if there are no rows at all.

Reply:
The test case given by the user is about exactly this, and
and there is another row in the partition already.
[#10081](https://github.com/scylladb/scylla/issues/10081)

But it's a good point that *when* to materialize a static row is
also a decision point.